### PR TITLE
Run tests against sqlite and postgres

### PR DESCRIPTION
**Pull request by shouichi**
This is the first step to tackle https://github.com/Shopify/maintenance_tasks/issues/462.

### CODE REVIEW

File: test/lib/generators/maintenance_tasks/install_generator_test.rb [hunk 1]
```diff
@@ -9,13 +9,20 @@ class InstallGeneratorTest < Rails::Generators::TestCase
     destination SAMPLE_APP_PATH
     setup :prepare_destination
 
+    original_adapter = ENV["DB"]
+
     setup do
+      # Force the sample app to use sqlite. This is because both the dummy app
+      # and the sample use the same postgres database, ended up migration
+      # failure.
+      ENV["DB"] = "sqlite"
```
Comment (shouichi) [inline test/lib/generators/maintenance_tasks/install_generator_test.rb:18]: This is the only tricky part that I failed to find a cleaner solution.

**Review by shouichi — COMMENTED** [review]
This PR essentially switches database adapters on database.yml and others are just adjustments except for the part I made a comment.

**Comment by adrianna-chang-shopify** [comment]
Hey @Shopify/rails, could I get thoughts / feedback on doing the above ^ to handle testing the install generator properly with PostgreSQL?

File: test/dummy/config/database.yml [hunk 2]
```diff
@@ -1,9 +1,6 @@
-# SQLite. Versions 3.8.0 and up are supported.
-#   gem install sqlite3
-#
-#   Ensure the SQLite 3 gem is defined in your Gemfile
-#   gem 'sqlite3'
-#
+<% adapter = ENV.fetch("DB") { "sqlite" } %>
+<% case adapter
```
[the remark this answers is not in the document]
Comment (shouichi) [reply test/dummy/config/database.yml:2]: Thanks for the suggestion. The `adapter` variable is used for raising an error when it's unknown, i.e.,

```
<% else %>
	<% raise "unknown adapter: #{adapter"} %>
<% end %>
```

I believe it's a good practice to raise an unknown adapter. Do you still want me to change?

File: test/dummy/config/sample_app_postgres_database.yml [hunk 3]
```diff
@@ -0,0 +1,14 @@
+# Database settings for postgres must be in sync with
```
Comment (shouichi) [inline test/dummy/config/sample_app_postgres_database.yml:1]: It turns out that rails 6.0 does not allow us to put yaml files in the fixtures directory https://github.com/rails/rails/pull/36884. It's a PR that @gmcgibbon sent, what a coincidence!

**Review by shouichi — COMMENTED** [review]
Hi, guys. Thanks for the help! I applied most of the suggestions and it __mostly__ works.

The last thing that needs to be fixed is `MaintenanceTasks::RunTest#test_#reload_status_does_not_use_query_cache` which probabilistically fails when postgres.

**Comment by shouichi** [comment]
Found the cause of `MaintenanceTasks::RunTest#test_#reload_status_does_not_use_query_cache` random failure.

When the test is executed first, active record first executes `SHOW search_path` to find out which postgres schemas to search. The solution I came up with is to run a query (e.g., `Run.first`) in setup and it worked. Is there any better solution?

File: test/dummy/config/database.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (adrianna-chang-shopify) [reply test/dummy/config/database.yml:2]: You can still store it in a variable (`<% case (adapter = ENV.fetch("DB")) %>`), but I agree with Gannon that it's more straightforward to set the default in the test helper instead.

**Comment by adrianna-chang-shopify** [comment]
@shouichi you can filter that query out from the `count_uncached_queries` method:
```ruby
    def count_uncached_queries(&block)
      count = 0

      query_cb = ->(*, payload) do
        unless payload[:cached] || payload[:sql].match?("SHOW search_path")
          count += 1
        end
      end
      ActiveSupport::Notifications.subscribed(query_cb,
        "sql.active_record",
        &block)

      count
    end
```

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -5,13 +5,28 @@ on: [push, pull_request]
 jobs:
   build:
     runs-on: ubuntu-latest
-    name: Ruby ${{ matrix.ruby }} / ${{ matrix.gemfile }}
+    name: Ruby ${{ matrix.ruby }} / ${{ matrix.gemfile }} / ${{ matrix.database }}
     strategy:
+      fail-fast: false
```
Comment (adrianna-chang-shopify) [inline .github/workflows/ci.yml:10]: Why this change?

**Comment by shouichi** [comment]
Fixed all the problems, please take a look. Feel free to ping me if I need to rebase or squash.

File: test/dummy/config/database.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (adrianna-chang-shopify) [reply test/dummy/config/database.yml:2]: Let's still set this default in the `test_helper` 😄

File: test/dummy/config/database.yml [hunk 5]
```diff
@@ -13,13 +10,36 @@ development:
...
 test:
   <<: *default
   database: db/test.sqlite3
 
 production:
   <<: *default
   database: db/production.sqlite3
+<% when "postgres" %>
+# Database settings for postgres must be in sync with
+# test/dummy/config/sample_app_postgres_database.yml.
```
Comment (adrianna-chang-shopify) [inline test/dummy/config/database.yml:22]: Let's ✂️  this comment

File: test/dummy/config/sample_app_postgres_database.yml [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+# test/dummy/config/database.yml.
```
Comment (adrianna-chang-shopify) [inline test/dummy/config/sample_app_postgres_database.yml:2]: Let's ✂️  this comment as well

**Review by adrianna-chang-shopify — APPROVED** [review]
Great work, thanks @shouichi ! A couple small comments. Let's also squash these commits before merging.

File: .github/workflows/ci.yml [hunk 7]
```diff
@@ -5,13 +5,27 @@ on: [push, pull_request]
...
     strategy:
       matrix:
         gemfile: [Gemfile, gemfiles/rails_6_0.gemfile, gemfiles/rails_edge.gemfile]
         ruby: [2.7, 3.0]
+        database: [sqlite, postgres]
     env:
       BUNDLE_GEMFILE: ${{ matrix.gemfile }}
+      DB: ${{ matrix.database }}
+    services:
+      postgres:
```
Comment (etiennebarrie) [inline .github/workflows/ci.yml:18]: Is there a way to have a service run conditionally? Better yet, the environment that runs CI already has postgres, so we shouldn't need an external service like this:
https://github.com/actions/virtual-environments/blob/main/images/linux/Ubuntu2004-README.md#databases

**Review by etiennebarrie — COMMENTED** [review]
> could I get thoughts / feedback on doing the above ^ to handle testing the install generator properly with PostgreSQL?

How about using `DATABASE_URL` which overrides `config/database.yml` entirely, just for the postgres cases in the matrix?

**Comment by adrianna-chang-shopify** [comment]
> How about using DATABASE_URL which overrides config/database.yml entirely, just for the postgres cases in the matrix?

I think we'll still run into the same problem -- if we use a db with the same name (`maintenance_tasks_test`), Postgres will complain that the `maintenance_tasks_runs` table exists already in the install generator test. We could use a `DATABASE_URL` instead of the additional db config file, but I believe we'd need to have a different database. We could use psql to create one in CI, but this makes things not-so-nice for local testing. 

Also looked into starting PostgreSQL directly on the Ubuntu VM, but couldn't quite get it -- not sure how to provide a password and / or specify no password. I tried
```bash
  sudo systemctl start postgresql.service
  psql -h localhost -U postgres -w --no-password 'create database maintenance_tasks_test;'
---
  sudo systemctl start postgresql.service
  PG_PASSWORD=postgres psql -h localhost -U postgres 'create database maintenance_tasks_test;'
---
  sudo systemctl start postgresql.service
  sudo -u postgres psql postgres -c 'create database maintenance_tasks_test;'
```

But no luck with any of them. I think it's acceptable to start postgres in another service. If you'd like to invest some more time into it @etiennebarrie let me know -- otherwise I'll ship this tmrw afternoon.

File: .github/workflows/ci.yml (hunk 7 above)
Comment (shouichi) [reply .github/workflows/ci.yml:18]: > Is there a way to have a service run conditionally?

As far as I can tell, there's no way to start a service conditionally. https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#jobsjob_idservices

The github guide (https://docs.github.com/en/actions/guides/creating-postgresql-service-containers) uses docker for postgres, so it shouldn't be that bad. Also, using containers gives us reproducibility and composability which may amortize the cost of spinning up extra service when sqlite testing.

**Comment by shouichi** [comment]
Thanks for the feedback. Speaking of the local development, I plan to open a follow-up PR to introduce docker based development environment. This way we don't need to set up postgres locally. Does this make sense?

**Comment by adrianna-chang-shopify** [comment]
Agree that `DATABASE_URL` is a more correct solution to this problem 👍  So to clarify @shouichi, let's do the following before shipping:
- Add a step to CI to export the `DATABASE_URL` if the db is postgres, as Étienne did in 3b8f99edfc1558b8eed3437245c67c877ccfb219 (although to me it's more readable to have it as `echo "DATABASE_URL=postgresql://localhost/maintenance_tasks_test" >> $GITHUB_ENV`)
- Remove the changes to `config/database.yml`
- Change the install generator to use the `DATABASE_URL` instead, as proposed in 631c89df519b7ace188f2350fa4a981e4b2c9863

Thanks for your patience! 🙏

**Comment by shouichi** [comment]
@adrianna-chang-shopify Thanks for the clarification. The `DATABASE_URL` solution is much cleaner, it's great! Changed based on the suggestions, rebased, and squashed. Please take a look.
